documents/coredocument.go

```diff
@@ -448,6 +465,56 @@ func (cd *CoreDocument) createProofs(fromZKTree bool, docType string, dataLeaves
 	}, nil
 }
 
+//CreateFingerprint generates a fingerprint for a Core Document
+func (cd *CoreDocument) CreateFingerprint() ([]byte, error) {
+	f := newFingerprint()
+	if len(cd.Document.Roles) == 0 {
+		return []byte{}, nil
+	}
+	if len(cd.Document.TransitionRules) == 0 {
+		return []byte{}, nil
+	}
+	f.Fingerprint.Roles = cd.Document.Roles
+	f.Fingerprint.TransitionRules = cd.Document.TransitionRules
```

Comment:
Roles shouldn't be copied since these roles might contain Roles belonging to ReadRules as well. First we should copy the transition rules and loop though each rule and collect the roles. These collected roles will be copied to fingerprint.roles